### ISSUE TITLE
consider using requests.sessions.Session

### ISSUE THREAD

**Issue by jetpackdanger** [attzonko/mmpy_bot#131]
I noticed while debugging one of our bots that mmpy_bot seems to make a fresh connection to the Mattermost server for each request. This slows things down, especially when TLS is in use, and can make more chatty bots seem to chug a little bit.

Please consider using [requests.sessions.Session](https://requests.readthedocs.io/en/master/user/advanced/#session-objects) instead of the requests module directly.  This would entail something like the diff below (untested, but I've used similar code in other projects).

It's almost possible to monkeypatch mmpy_bot to do this, but `MattermostClient.__init__` refers to `requests.packages` in a couple of places, and this import is of course not included in the Session object's API. (This only impacts `ssl_verify=False`, which we don't use anyway.)

```diff
diff --git a/mmpy_bot/mattermost.py b/mmpy_bot/mattermost.py
index 7020bc4..d09536e 100644
--- a/mmpy_bot/mattermost.py
+++ b/mmpy_bot/mattermost.py
@@ -11,6 +11,7 @@ logger = logging.getLogger(__name__)

 class MattermostAPI(object):
     def __init__(self, url, ssl_verify, token):
+        self.session = request.sessions.Session()
         self.url = url
         self.token = token
         self.initial = None
@@ -62,7 +63,7 @@ class MattermostAPI(object):

     def get(self, request):
         return json.loads(
-            requests.get(
+            self.session.get(
                 self.url + request,
                 headers=self._get_headers(),
                 verify=self.ssl_verify
@@ -108,7 +109,7 @@ class MattermostAPI(object):
                    parse=None, link_names=None, attachments=None,
                    unfurl_links=None, unfurl_media=None, icon_url=None,
                    icon_emoji=None, ssl_verify=True, **kwargs):
-        return requests.post(
+        return self.session.post(
             url, data={
                 'payload': json.dumps({
                     'channel': channel,
@@ -142,7 +143,7 @@ class MattermostAPI(object):
                 response.raise_for_status()

     def _login(self, props):
-        return requests.post(
+        return self.session.post(
             self.url + '/users/login',
             data=json.dumps(props),
             verify=self.ssl_verify,
@@ -165,7 +166,7 @@ class MattermostAPI(object):
         return self.get('/users/me')

     def post(self, request, data):
-        return json.loads(requests.post(
+        return json.loads(self.session.post(
             self.url + request,
             headers=self._get_headers(),
             data=json.dumps(data),
@@ -173,14 +174,14 @@ class MattermostAPI(object):
         ).text)

     def delete(self, request):
-        return json.loads(requests.delete(
+        return json.loads(self.session.delete(
             self.url + request,
             headers=self._get_headers(),
             verify=self.ssl_verify
         ).text)

     def put(self, request, data):
-        return json.loads(requests.put(
+        return json.loads(self.session.put(
             self.url + request,
             headers=self._get_headers(),
             data=json.dumps(data),
@@ -203,7 +204,7 @@ class MattermostAPI(object):
             'files': file,
             'channel_id': (None, channel_id)
         }
-        return json.loads(requests.post(
+        return json.loads(self.session.post(
             self.url + '/files',
             headers=self._get_headers(),
             files=files,
```

**Comment by attzonko** [comment]
I can see how this would be helpful. The use of requests in init is really not needed. Why don't you make a PR and test it?

**Comment by Tobsic** [comment]
Another improvement would be to set `headers` in the session, after login. So it would be easy to use the session to make all kinds of request (just stumbled over the missing `delete` method, required to delete a post)
But #148 would make this obsolete as well ;)

**Comment by jneeven** [comment]
Closed by #153, which now uses `mattermost-driver` as the connection interface (and therefore does not have this problem :))
